baseplate/sidecars/event_publisher.py

```diff
@@ -1,3 +1,13 @@
+# this is here because we want the gevent monkeypatching to happen before
+# any pieces of baseplate are imported at all, as would happen if we were
+# to have this code be in the baseplate package.
+from gevent.monkey import patch_all
+
+from baseplate.server.monkey import patch_stdlib_queues
+
+patch_all()
+patch_stdlib_queues()
```

Comment:
Why is monkeypatching required. What issue does Baseplate have if no monkey patching happens.

Reply:
I'm not convinced it is _required_, and would believe that there is a better way to do it. My understanding is that `gevent` allows concurrent execution. The monkey patching seems to make a subset of the standard library (which would normally be blocking) work concurrently https://www.gevent.org/api/gevent.monkey.html

Without this monkey patching, the server would be set up properly but all requests to it would hang indefinitely because other execution was happening instead (and then suddenly when the process was interrupted they would complete). After the monkey patching, the requests were handled concurrently as they came in.

Reply:
I'm actually surprised that sidecars are launched by loading this module directly. gevent loops work similar to javascript, so what you're describing is that something is in the event loop is adding more to where it's executing than the end (likely the [while loop here](https://github.com/reddit/baseplate.py/blob/BASE-122/baseplate/sidecars/event_publisher.py#L240)), which gevent solves by patching sockets to be in the middle (which lets it do other things while waiting for a response to socket events...I think?).

This is a bit too magic for me and I think the underlying issue is the while loop as it stands. I'd rather not rely on sockets to force this to happen. This can be resolved by `gevent.sleep(0)` within the loop to ensure gevent gets a chance to run other stuff in the event loop.

Reply:
> I'm actually surprised that sidecars are launched by loading this module directly

That's fair, it could definitely make sense to make a separate entrypoint. But yeah, that's how it currently works https://github.snooguts.net/reddit/docker-baseplate-sidecars/blob/master/entrypoint.sh

> gevent.sleep(0)

Interesting, this makes sense to me though it does not work as-is. Still working on understanding how it differs from what I expected

Reply:
To clarify Nate's comment a bit (since I didn't fully understand it), the point was that _only_ monkey patching just works because there happens to be things with socket events inside of that while loop. Rather than relying on that, better to be explicit and on top of monkey patching, also add the sleep.